File: backend/app/schemas/similarity.py
"""Public camel-case API contracts for Phase 9 similarity."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, field_validator

from app.models.similarity_enums import (
    ConsistencyStatus,
    SimilarityAnalysisStatus,
    SimilarityCategory,
    SimilarityJobStatus,
    SimilarityJobType,
    SimilarityRunStatus,
)
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData


def _required_reason(value: object) -> object:
    if not isinstance(value, str) or not value.strip():
        raise ValueError("Reason is required.")
    return value.strip()


class SimilarityStartRequest(ApiSchema):
    document_file_id: UUID
    compliance_run_id: UUID | None = None
    language_detection_run_id: UUID | None = None
    force: bool = False


class SimilarityRerunRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=2000)

    _reason = field_validator("reason", mode="before")(_required_reason)


class SimilarityDocumentReference(ApiSchema):
    id: UUID
    base_document_code: str
    title: str
    department_id: UUID


class SimilarityRevisionReference(ApiSchema):
    id: UUID
    revision_code: str
    full_document_code: str


class SimilarityFileReference(ApiSchema):
    id: UUID
    filename: str
    file_extension: str


class SimilarityRequesterReference(ApiSchema):
    id: UUID
    name: str


class SimilarityQueuedResponse(ApiSchema):
    id: UUID
    status: SimilarityJobStatus
    progress: int = Field(ge=0, le=100)
    document_file_id: UUID
    run_id: UUID | None = None
    reused_existing_result: bool = False


class SimilarityJobResponse(ApiSchema):
    id: UUID
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    compliance_run_id: UUID
    language_detection_run_id: UUID
    document: SimilarityDocumentReference | None = None
    revision: SimilarityRevisionReference | None = None
    file: SimilarityFileReference | None = None
    job_type: SimilarityJobType
    status: SimilarityJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None = None
    source_content_hash: str | None = None
    provider: str
    model_name: str
    requested_by: SimilarityRequesterReference | None = None
    requested_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    failed_at: datetime | None = None
    cancelled_at: datetime | None = None
    attempt_number: int = Field(ge=1)
    maximum_attempts: int = Field(ge=1)
    error_code: str | None = None
    error_message: str | None = None
    error_details: dict[str, Any] | None = None
    result_summary: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class SimilarityCancelResponse(ApiSchema):
    id: UUID
    status: SimilarityJobStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None = None
    cancelled_at: datetime | None = None


class SimilarityJobListResponse(PaginationData[SimilarityJobResponse]):
    pass


class SimilarityRunResponse(ApiSchema):
    id: UUID
    similarity_job_id: UUID
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    compliance_run_id: UUID
    language_detection_run_id: UUID
    document: SimilarityDocumentReference | None = None
    revision: SimilarityRevisionReference | None = None
    file: SimilarityFileReference | None = None
    provider: str
    model_name: str
    model_version: str | None = None
    status: SimilarityRunStatus
    source_content_hash: str
    translation_group_count: int = Field(ge=0)
    eligible_group_count: int = Field(ge=0)
    analysed_group_count: int = Field(ge=0)
    skipped_group_count: int = Field(ge=0)
    failed_group_count: int = Field(ge=0)
    average_similarity: float | None = Field(default=None, ge=0, le=1)
    minimum_similarity: float | None = Field(default=None, ge=0, le=1)
    maximum_similarity: float | None = Field(default=None, ge=0, le=1)
    id_en_average_similarity: float | None = Field(
        default=None, ge=0, le=1
    )
    id_zh_average_similarity: float | None = Field(
        default=None, ge=0, le=1
    )
    en_zh_average_similarity: float | None = Field(
        default=None, ge=0, le=1
    )
    high_similarity_groups: int = Field(ge=0)
    review_similarity_groups: int = Field(ge=0)
    low_similarity_groups: int = Field(ge=0)
    unavailable_similarity_groups: int = Field(ge=0)
    number_mismatch_count: int = Field(ge=0)
    date_mismatch_count: int = Field(ge=0)
    measurement_mismatch_count: int = Field(ge=0)
    reference_mismatch_count: int = Field(ge=0)
    negation_mismatch_count: int = Field(ge=0)
    warnings: list[str] | list[dict[str, Any]] = Field(default_factory=list)
    metrics: dict[str, Any] = Field(default_factory=dict)
    started_at: datetime
    completed_at: datetime | None = None
    requested_by: SimilarityRequesterReference | None = None
    created_at: datetime


class SimilarityRunListResponse(PaginationData[SimilarityRunResponse]):
    pass


class SimilaritySummaryResponse(ApiSchema):
    run_id: UUID
    status: SimilarityRunStatus
    average_similarity: float | None = Field(default=None, ge=0, le=1)
    minimum_similarity: float | None = Field(default=None, ge=0, le=1)
    maximum_similarity: float | None = Field(default=None, ge=0, le=1)
    translation_group_count: int = Field(ge=0)
    eligible_group_count: int = Field(ge=0)
    analysed_group_count: int = Field(ge=0)
    skipped_group_count: int = Field(ge=0)
    failed_group_count: int = Field(ge=0)
    categories: dict[str, int] = Field(default_factory=dict)
    pair_averages: dict[str, float | None] = Field(default_factory=dict)
    mismatches: dict[str, int] = Field(default_factory=dict)
    section_count: int = Field(ge=0)
    finding_count: int = Field(default=0, ge=0)
    warnings: list[str] | list[dict[str, Any]] = Field(default_factory=list)


class TranslationSimilarityResultResponse(ApiSchema):
    id: UUID
    similarity_run_id: UUID
    translation_group_id: UUID
    detected_section_id: UUID | None = None
    container_id: UUID | None = None
    source_reference: str
    source_language_code: str
    target_language_code: str
    source_member_id: UUID | None = None
    target_member_id: UUID | None = None
    source_text_hash: str
    target_text_hash: str
    source_text_snippet: str | None = None
    target_text_snippet: str | None = None
    similarity_score: float | None = Field(default=None, ge=0, le=1)
    similarity_category: SimilarityCategory
    confidence: float = Field(ge=0, le=1)
    structural_group_confidence: float | None = Field(
        default=None,
        ge=0,
        le=1,
    )
    ocr_confidence: float | None = Field(default=None, ge=0, le=1)
    analysis_status: SimilarityAnalysisStatus
    source_character_count: int = Field(ge=0)
    target_character_count: int = Field(ge=0)
    length_ratio: float | None = Field(default=None, ge=0)
    number_consistency_status: ConsistencyStatus
    date_consistency_status: ConsistencyStatus
    measurement_consistency_status: ConsistencyStatus
    reference_consistency_status: ConsistencyStatus
    negation_consistency_status: ConsistencyStatus
    number_details: dict[str, Any] = Field(default_factory=dict)
    date_details: dict[str, Any] = Field(default_factory=dict)
    measurement_details: dict[str, Any] = Field(default_factory=dict)
    reference_details: dict[str, Any] = Field(default_factory=dict)
    negation_details: dict[str, Any] = Field(default_factory=dict)
    chunk_count_source: int = Field(ge=0)
    chunk_count_target: int = Field(ge=0)
    metrics: dict[str, Any] = Field(default_factory=dict)
    warnings: list[str] | list[dict[str, Any]] = Field(default_factory=list)
    finding_count: int = Field(default=0, ge=0)
    related_finding_ids: list[UUID] = Field(default_factory=list)
    created_at: datetime


class TranslationSimilarityResultListResponse(
    PaginationData[TranslationSimilarityResultResponse]
):
    pass


class SectionSimilaritySummaryResponse(ApiSchema):
    id: UUID
    similarity_run_id: UUID
    detected_section_id: UUID | None = None
    canonical_section_code: str
    total_groups: int = Field(ge=0)
    eligible_groups: int = Field(ge=0)
    analysed_groups: int = Field(ge=0)
    average_similarity: float | None = Field(default=None, ge=0, le=1)
    minimum_similarity: float | None = Field(default=None, ge=0, le=1)
    low_similarity_groups: int = Field(ge=0)
    number_mismatches: int = Field(ge=0)
    date_mismatches: int = Field(ge=0)
    measurement_mismatches: int = Field(ge=0)
    reference_mismatches: int = Field(ge=0)
    negation_mismatches: int = Field(ge=0)
    pairwise_summary: dict[str, Any] = Field(default_factory=dict)
    metrics: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class SectionSimilaritySummaryListResponse(
    PaginationData[SectionSimilaritySummaryResponse]
):
    pass

File: frontend/src/types/similarity.ts
import type { DocumentReference, DocumentUserSummary } from './document';
import type { PaginatedData } from './masterData';

export const similarityCategories = [
  'HIGH',
  'ACCEPTABLE',
  'NEEDS_REVIEW',
  'LOW',
  'NOT_EVALUATED',
] as const;

export type SimilarityCategory = (typeof similarityCategories)[number];

export const similarityJobStatuses = [
  'QUEUED',
  'LOADING_CONTEXT',
  'LOADING_MODEL',
  'ALIGNING_GROUPS',
  'ENCODING',
  'CALCULATING_SIMILARITY',
  'CHECKING_CONSISTENCY',
  'AGGREGATING',
  'GENERATING_FINDINGS',
  'PERSISTING',
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCEL_REQUESTED',
  'CANCELLED',
] as const;

export type SimilarityJobStatus = (typeof similarityJobStatuses)[number];
export type SimilarityJobType =
  'INITIAL_SIMILARITY' | 'REANALYSIS' | 'MANUAL_SIMILARITY';
export type SimilarityRunStatus = 'COMPLETED' | 'PARTIALLY_COMPLETED' | 'FAILED';
export type SimilarityResultStatus =
  | 'COMPLETED'
  | 'SKIPPED_TOO_SHORT'
  | 'SKIPPED_UNSUPPORTED'
  | 'FAILED'
  | 'INSUFFICIENT_CONTENT';
export type ConsistencyStatus =
  | 'MATCH'
  | 'MISMATCH'
  | 'AMBIGUOUS'
  | 'NOT_APPLICABLE'
  | 'POTENTIALLY_EQUIVALENT'
  | 'POSSIBLE_NEGATION_MISMATCH';
export type SupportedLanguageCode = 'id' | 'en' | 'zh';
export type SimilarityExportFormat = 'json' | 'xlsx';

export const terminalSimilarityJobStatuses: readonly SimilarityJobStatus[] = [
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCELLED',
];

export const isTerminalSimilarityJobStatus = (status: SimilarityJobStatus): boolean =>
  terminalSimilarityJobStatuses.includes(status);

export const isActiveSimilarityJobStatus = (status: SimilarityJobStatus): boolean =>
  !isTerminalSimilarityJobStatus(status);

export interface SimilarityStartRequest {
  documentFileId: string;
  complianceRunId?: string | null;
  languageDetectionRunId?: string | null;
  force?: boolean;
}

export interface SimilarityRerunRequest {
  reason: string;
}

export interface SimilarityResultSummary {
  runId?: string | null;
  averageSimilarity?: number | null;
  minimumSimilarity?: number | null;
  translationGroups?: number;
  lowGroups?: number;
  numberMismatches?: number;
  dateMismatches?: number;
  measurementMismatches?: number;
  referenceMismatches?: number;
  negationMismatches?: number;
  analysedGroups?: number;
  lowSimilarityGroups?: number;
  findingCount?: number;
}

export interface SimilarityJob {
  id: string;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  complianceRunId: string | null;
  languageDetectionRunId: string | null;
  jobType: SimilarityJobType;
  status: SimilarityJobStatus;
  progress: number;
  currentStage: string | null;
  provider: string;
  providerName?: string | null;
  modelName: string;
  sourceContentHash: string | null;
  attemptNumber: number;
  maximumAttempts: number;
  retryCount?: number;
  requestedBy: DocumentUserSummary | null;
  requestedAt: string;
  startedAt: string | null;
  completedAt: string | null;
  failedAt: string | null;
  cancelledAt: string | null;
  errorCode: string | null;
  errorMessage: string | null;
  errorDetails: Readonly<Record<string, unknown>> | null;
  resultSummary: SimilarityResultSummary | null;
  createdAt: string;
  updatedAt: string;
  document?: {
    id: string;
    baseDocumentCode: string;
    title: string;
    departmentId: string;
  } | null;
  revision?: {
    id: string;
    revisionCode: string;
    fullDocumentCode: string;
  } | null;
  file?: {
    id: string;
    filename: string;
    fileExtension: string;
  } | null;
}

export interface SimilarityJobListParams {
  page: number;
  pageSize: number;
  status?: SimilarityJobStatus | readonly SimilarityJobStatus[];
  departmentId?: string;
  search?: string;
  requestedFrom?: string;
  requestedTo?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
}

export type SimilarityJobList = PaginatedData<SimilarityJob>;

export interface SimilarityQueuedResult {
  id: string;
  jobId?: string;
  status: SimilarityJobStatus;
  progress: number;
  documentFileId: string;
  runId: string | null;
  reusedExistingResult: boolean;
  message?: string;
}

export interface SimilarityCancelResult {
  id: string;
  status: SimilarityJobStatus;
  progress: number;
  currentStage: string | null;
  cancelledAt: string | null;
}

export interface LanguagePairSummary {
  sourceLanguage: SupportedLanguageCode;
  targetLanguage: SupportedLanguageCode;
  averageSimilarity: number | null;
  minimumSimilarity: number | null;
  groupsAnalysed: number;
  high: number;
  acceptable: number;
  needsReview: number;
  low: number;
  notEvaluated: number;
}

export interface SimilarityRun {
  id: string;
  similarityJobId: string | null;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  complianceRunId: string | null;
  languageDetectionRunId: string | null;
  providerName?: string;
  provider: string;
  modelName: string;
  modelVersion: string | null;
  status: SimilarityRunStatus;
  sourceContentHash: string;
  translationGroupCount: number;
  eligibleGroupCount: number;
  analysedGroupCount: number;
  skippedGroupCount: number;
  failedGroupCount: number;
  totalGroups?: number;
  evaluatedGroups?: number;
  skippedGroups?: number;
  failedGroups?: number;
  averageSimilarity: number | null;
  minimumSimilarity: number | null;
  maximumSimilarity: number | null;
  idEnAverageSimilarity: number | null;
  idZhAverageSimilarity: number | null;
  enZhAverageSimilarity: number | null;
  highSimilarityGroups: number;
  reviewSimilarityGroups: number;
  lowSimilarityGroups: number;
  unavailableSimilarityGroups: number;
  highCount?: number;
  acceptableCount?: number;
  needsReviewCount?: number;
  lowCount?: number;
  notEvaluatedCount?: number;
  numberMismatchCount: number;
  dateMismatchCount: number;
  measurementMismatchCount: number;
  referenceMismatchCount: number;
  negationMismatchCount: number;
  languagePairSummaries?: readonly LanguagePairSummary[];
  warnings: readonly (string | Readonly<Record<string, unknown>>)[];
  metrics: Readonly<Record<string, unknown>>;
  requestedBy: DocumentUserSummary | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string;
  document?: {
    id: string;
    baseDocumentCode: string;
    title: string;
    departmentId: string;
  } | null;
  revision?: {
    id: string;
    revisionCode: string;
    fullDocumentCode: string;
  } | null;
  file?: {
    id: string;
    filename: string;
    fileExtension: string;
  } | null;
}

export interface SimilaritySummary {
  runId: string;
  status: SimilarityRunStatus;
  averageSimilarity: number | null;
  minimumSimilarity: number | null;
  maximumSimilarity: number | null;
  translationGroupCount: number;
  eligibleGroupCount: number;
  analysedGroupCount: number;
  skippedGroupCount: number;
  failedGroupCount: number;
  categories: Readonly<Record<string, number>>;
  pairAverages: Readonly<Record<string, number | null>>;
  mismatches: Readonly<Record<string, number>>;
  sectionCount: number;
  groupsAnalysed?: number;
  high?: number;
  acceptable?: number;
  needsReview?: number;
  low?: number;
  notEvaluated?: number;
  numberMismatches?: number;
  dateMismatches?: number;
  measurementMismatches?: number;
  referenceMismatches?: number;
  negationMismatches?: number;
  findingCount: number;
  qualityStatus?: QualityStatus;
  warnings: readonly (string | Readonly<Record<string, unknown>>)[];
  languagePairs?: readonly LanguagePairSummary[];
}

export type QualityStatus =
  'HIGH_QUALITY' | 'ACCEPTABLE' | 'NEEDS_REVIEW' | 'LOW_QUALITY' | 'NOT_EVALUATED';

export interface ConsistencyDetail {
  sourceValues: readonly string[];
  targetValues: readonly string[];
  missingFromSource?: readonly string[];
  missingFromTarget?: readonly string[];
  warning?: string | null;
}

export interface TranslationSimilarityResult {
  id: string;
  similarityRunId: string;
  translationGroupId: string | null;
  detectedSectionId: string | null;
  sectionCode?: string | null;
  sectionName?: string | null;
  containerId: string | null;
  sourceReference: string | null;
  groupLabel?: string | null;
  sourceLanguageCode: SupportedLanguageCode;
  targetLanguageCode: SupportedLanguageCode;
  sourceLanguage?: SupportedLanguageCode;
  targetLanguage?: SupportedLanguageCode;
  sourceMemberId: string | null;
  targetMemberId: string | null;
  sourceTextSnippet: string | null;
  targetTextSnippet: string | null;
  sourceTextHash: string;
  targetTextHash: string;
  similarityScore: number | null;
  similarityCategory: SimilarityCategory;
  confidence: number;
  confidenceScore?: number | null;
  structuralGroupConfidence?: number | null;
  ocrConfidence?: number | null;
  analysisStatus: SimilarityResultStatus;
  status?: SimilarityResultStatus;
  sourceCharacterCount: number;
  targetCharacterCount: number;
  lengthRatio: number | null;
  lengthRatioStatus?: ConsistencyStatus;
  numberConsistencyStatus: ConsistencyStatus;
  numberStatus?: ConsistencyStatus;
  numberDetails: Readonly<Record<string, unknown>>;
  dateConsistencyStatus: ConsistencyStatus;
  dateStatus?: ConsistencyStatus;
  dateDetails: Readonly<Record<string, unknown>>;
  measurementConsistencyStatus: ConsistencyStatus;
  measurementStatus?: ConsistencyStatus;
  measurementDetails: Readonly<Record<string, unknown>>;
  referenceConsistencyStatus: ConsistencyStatus;
  referenceStatus?: ConsistencyStatus;
  referenceDetails: Readonly<Record<string, unknown>>;
  negationConsistencyStatus: ConsistencyStatus;
  negationStatus?: ConsistencyStatus;
  negationDetails: Readonly<Record<string, unknown>>;
  chunkCountSource: number;
  chunkCountTarget: number;
  chunkCount?: number;
  findingCount: number;
  relatedFindingIds: readonly string[];
  metrics: Readonly<Record<string, unknown>>;
  warnings: readonly (string | Readonly<Record<string, unknown>>)[];
  createdAt: string;
}

export interface SimilarityResultListParams {
  sectionId?: string;
  sourceLanguage?: SupportedLanguageCode;
  targetLanguage?: SupportedLanguageCode;
  similarityCategory?: SimilarityCategory;
  minimumScore?: number;
  maximumScore?: number;
  hasNumberMismatch?: boolean;
  hasDateMismatch?: boolean;
  hasMeasurementMismatch?: boolean;
  hasReferenceMismatch?: boolean;
  hasNegationMismatch?: boolean;
  findingSeverity?: 'CRITICAL' | 'MAJOR' | 'MINOR' | 'INFO';
  search?: string;
  page: number;
  pageSize: number;
}

export type SimilarityResultList = PaginatedData<TranslationSimilarityResult>;

export interface SectionSimilaritySummary {
  id: string;
  similarityRunId: string;
  detectedSectionId: string | null;
  canonicalSectionCode: string;
  sectionCode?: string | null;
  sectionName?: string | null;
  sectionOrder?: number | null;
  totalGroups: number;
  eligibleGroups: number;
  analysedGroups: number;
  evaluatedGroups?: number;
  averageSimilarity: number | null;
  minimumSimilarity: number | null;
  lowSimilarityGroups: number;
  numberMismatches: number;
  dateMismatches: number;
  measurementMismatches: number;
  referenceMismatches: number;
  negationMismatches: number;
  highCount?: number;
  acceptableCount?: number;
  needsReviewCount?: number;
  lowCount?: number;
  notEvaluatedCount?: number;
  mismatchCount?: number;
  pairwiseSummary: Readonly<Record<string, unknown>>;
  metrics: Readonly<Record<string, unknown>>;
  createdAt: string;
}

export type SectionSimilarityList =
  readonly SectionSimilaritySummary[] | PaginatedData<SectionSimilaritySummary>;

export interface SimilarityHistoryParams {
  page: number;
  pageSize: number;
}

export type SimilarityHistory = PaginatedData<SimilarityRun>;

export interface SimilarityDownload {
  blob: Blob;
  fileName: string | null;
}

export interface SimilarityDocumentReference {
  document: DocumentReference | null;
  revision: DocumentReference | null;
}
